### PR TITLE
feat(connector): [Adyen] Add pix support for adyen

### PR DESCRIPTION
**Pull request by Sangamesh26**
## Type of Change

- [x] New feature

## Description

Add support for `Pix` bank transfer through adyen

Api contract for `Pix`:
```
    "payment_method": "bank_transfer",
    "payment_method_type": "pix",
    "payment_method_data": {
        "bank_transfer": {
            "pix": {}
        }
    }
```

Pix sends `QrCode` data along with `Qr url`, QrCode data is converted into image data source is sent in next action.

Response that SDK can use :
```
  "next_action": {
        "type": "qr_code_information",
        "image_data_url": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAQgAAAEICAAAAACGnTUjAAADfUlEQVR4nO3QS25dOQwA0c7+F90foGpCgKCeE3cGqQN4QEq6fqgff/+V/xQChUAhUAgUAoVAIVAIFAKFQCFQCBQChUAhUAgUAoVAIVAIFAKFQCFQCBQChUAhUAgUAoVAIXCG+PHv31f4Xd87y708n3td5xffbwqBQqAQKAQKgUKgEHgOcd3Tdn/uneV+8p7nc7683i8ECoFCoBAoBAqBQuDjEM7TPHee5vmc9bp3nua586YQKAQKgUKgECgECoFvD+G88d4033nP/Zynee68KQQKgUKgECgECoFC4H8PMffOcj95z/M5T/PceVMIFAKFQCFQCBQChcDHIS7zvvPF+5PvPZ/z5fV+IVAIFAKFQCFQCBQCzyE+5Xd9/6vnT/l+UwgUAoVAIVAIFAKFwBniZ/nD/T/bLPfyfO5/tUKgECgECoFCoBAoBM4Q1w+Z584b72ne99z9nC/el+/mfioECoFCoBAoBAqBQuAMcfEfye/Nva7zje8mvzPPt/2mECgECoFCoBAoBAqBj0P4Dya/4/mcde3l+dx/l0KgECgECoFCoBAoBM4Q/qCN773nrGsvz907b7Z72/5SCBQChUAhUAgUAoXAcwjvOV+8r/nuOpf3vnou720KgUKgECgECoFCoBB4DqHtvvc832bNvbPcy3P3c5b7TxUChUAhUAgUAoVAIXCG2PgD5vvXvfMr3+l6P+9fCoFCoBAoBAqBQqAQ+HIIzR80v+f5tb/myXN5b9tfCoFCoBAoBAqBQqAQOEPMD2t7N+/Pe567d5Z7zfOL7+c795tCoBAoBAqBQqAQKATOED/LH3T9H+9ttve+m+fuNc+nQqAQKAQKgUKgECgEzhDzg6/8ru+dX/lOvnf/Or8qBAqBQqAQKAQKgULgOcR1T9v91/01T/PcefJ8UwgUAoVAIVAIFAKFwMchnKd5fs36dH/xnXw/91MhUAgUAoVAIVAIFALfHkLu5bl758nzzeu77Z4KgUKgECgECoFCoBD4bSEuvp/v5v51vhQChUAhUAgUAoVAIfBxiMvr/Xlvzpp752k7d38pBAqBQqAQKAQKgULgOcSn/K7vr1lzP+eN9zTvz/OpECgECoFCoBAoBAqBM8SfohAoBAqBQqAQKAQKgUKgECgECoFCoBAoBAqBQqAQKAQKgUKgECgECoFCoBAoBAqBQqAQKAQKgX8A/Ak4Wyhn58sAAAAASUVORK5CYII=",
        "qr_code_url": "https://test.adyen.com/hpp/generateQRCodeImage.shtml?url=TestQRCodeEMVToken"
    },
```

### Additional Changes

- [x] This PR modifies the API contract

## Motivation and Context

Add support for new bank transfers

## How did you test it?

Screen Shot 2023-07-13 at 2 28 41 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code

### CODE REVIEW

File: crates/router/src/connector/adyen/transformers.rs [hunk 1]
```diff
@@ -1805,32 +1868,76 @@ pub fn get_redirection_response(
...
+            services::RedirectForm::Form {
+                endpoint: url.to_string(),
+                method: response
+                    .action
+                    .method
+                    .to_owned()
+                    .unwrap_or(services::Method::Get),
+                form_fields,
+            }
+        }),
```
Comment (jagan-jaya) [inline crates/router/src/connector/adyen/transformers.rs:1889]:
```suggestion
        _ => response.action.url.to_owned().map(|url| services::RedirectForm::from((url, services::Method::Get))),
```

File: crates/router/src/connector/adyen/transformers.rs [hunk 2]
```diff
@@ -1805,32 +1868,76 @@ pub fn get_redirection_response(
...
     };
+
     Ok((status, error, payments_response_data))
 }
 
+pub fn get_connector_metadata(
+    response: AdyenNextActionResponse,
+) -> errors::CustomResult<Option<serde_json::Value>, errors::ConnectorError> {
+    let connector_metadata = match response.action.type_of_response {
+        ActionType::QrCode => {
```
Comment (jagan-jaya) [inline crates/router/src/connector/adyen/transformers.rs:1911]: try to create an utility like this ```services::QrCode::from(.qr_code_data)``` and remove entire code in this function

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

**Review by jagan-jaya — APPROVED** [review]

File: crates/router/src/core/payments/transformers.rs [hunk 3]
```diff
@@ -345,6 +345,7 @@ where
 
                 if payment_intent.status == enums::IntentStatus::RequiresCustomerAction
                     || bank_transfer_next_steps.is_some()
+                    || next_action_containing_qr_code.is_some()
```
[the remark this answers is not in the document]
Comment (Sangamesh26) [reply crates/router/src/core/payments/transformers.rs:348]: sure

File: crates/router/src/connector/adyen/transformers.rs [hunk 4]
```diff
@@ -1641,6 +1608,52 @@ impl<'a>
...
+            amount,
+            merchant_account: auth_type.merchant_account,
+            payment_method,
+            reference: item.payment_id.to_string(),
+            return_url,
+            browser_info: None,
+            shopper_interaction,
+            recurring_processing_model: None,
+            additional_data: None,
+            shopper_name: None,
```
Comment (ArjunKarthik) [inline crates/router/src/connector/adyen/transformers.rs:1641]: Whatever we are passing is different from the docs expectation, Let's have discussion on this tomorrow

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/adyen/transformers.rs (hunk 4 above)
Comment (Sangamesh26) [reply crates/router/src/connector/adyen/transformers.rs:1641]: All other attributes in the adyen docs is not compulsory, so only the required fields are included.

File: crates/router/src/connector/adyen/transformers.rs [hunk 5]
```diff
@@ -400,16 +389,11 @@ pub struct SamsungPayPmData {
 }
 
 #[derive(Debug, Clone, Serialize)]
-pub struct PayBrightData {
+pub struct PmdForPaymentType {
```
[the remark this answers is not in the document]
Comment (Sangamesh26) [reply crates/router/src/connector/adyen/transformers.rs:392]: Will take up this in next PR

**Review by ArjunKarthik — APPROVED** [review]

**Review by ashokkjag — APPROVED** [review]

**Review by Abhicodes-crypto — APPROVED** [review]

**Pull request merged**